File: native/android/AndroidAwnCore/src/main/kotlin/me/carda/android_awn_core/NotificationBuilder.kt
package me.carda.android_awn_core

import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.RemoteInput
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Builds notifications from a serialized `NotificationModel`, injects the full
 * model into the tap/dismiss PendingIntents so it can be recovered later, and
 * stamps the lifecycle events (created/displayed/pressed/dismissed).
 *
 * Created via a factory (an instance with injected helpers — not static utility
 * methods), mirroring the original AndroidAwnCore `NotificationBuilder`.
 */
class NotificationBuilder private constructor(
    private val stringUtils: StringUtils,
    private val mapUtils: MapUtils,
    private val jsonUtils: JsonUtils,
    private val bitmapUtils: BitmapUtils
) {
    companion object {
        fun getNewBuilder(): NotificationBuilder =
            NotificationBuilder(
                StringUtils.getInstance(),
                MapUtils.getInstance(),
                JsonUtils.getInstance(),
                BitmapUtils.getInstance()
            )
    }

    // MARK: - Build

    /** Builds an Android notification from a serialized model, or null if invalid. */
    fun build(context: Context, model: Map<String, Any?>): Notification? {
        val content = contentMap(model)
        val id = mapUtils.getInt(content[Definitions.NOTIFICATION_ID]) ?: return null
        val channelKey =
            mapUtils.getString(content[Definitions.NOTIFICATION_CHANNEL_KEY]) ?: return null

        val builder = NotificationCompat.Builder(context, channelKey)
            .setSmallIcon(context.applicationInfo.icon)
            .setAutoCancel(shouldAutoDismiss(content))
            .setContentIntent(contentIntent(context, id, model))
            .setDeleteIntent(deleteIntent(context, id, model))

        val title = mapUtils.getString(content[Definitions.NOTIFICATION_TITLE])
        if (!stringUtils.isNullOrEmpty(title)) builder.setContentTitle(title)
        val body = mapUtils.getString(content[Definitions.NOTIFICATION_BODY])
        if (!stringUtils.isNullOrEmpty(body)) builder.setContentText(body)

        applyImages(context, content, builder)
        applyActionButtons(context, id, model, builder)

        return builder.build()
    }

    // MARK: - Action buttons

    @Suppress("UNCHECKED_CAST")
    fun actionButtons(model: Map<String, Any?>): List<Map<String, Any?>> =
        (model[Definitions.NOTIFICATION_BUTTONS] as? List<Map<String, Any?>>) ?: emptyList()

    /** Adds an [NotificationCompat.Action] for each serialized action button. */
    private fun applyActionButtons(
        context: Context,
        id: Int,
        model: Map<String, Any?>,
        builder: NotificationCompat.Builder
    ) {
        actionButtons(model).forEachIndexed { index, button ->
            val key = mapUtils.getString(button[Definitions.NOTIFICATION_BUTTON_KEY])
            if (key.isNullOrEmpty()) return@forEachIndexed
            if (mapUtils.getBool(button[Definitions.NOTIFICATION_BUTTON_ENABLED]) == false) {
                return@forEachIndexed
            }

            val label = mapUtils.getString(button[Definitions.NOTIFICATION_BUTTON_LABEL]) ?: key
            val actionType =
                mapUtils.getString(button[Definitions.NOTIFICATION_ACTION_TYPE]) ?: "Default"
            val requireInput =
                mapUtils.getBool(button[Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT]) ?: false
            val isForeground = actionType.endsWith("Default")

            val pending = buttonPendingIntent(
                context, id, index, key, model, isForeground, requireInput
            )
            val actionBuilder = NotificationCompat.Action.Builder(
                buttonIcon(context, button), label, pending
            )
            if (requireInput) {
                actionBuilder.addRemoteInput(
                    RemoteInput.Builder(Definitions.NOTIFICATION_BUTTON_KEY_INPUT)
                        .setLabel(label)
                        .build()
                )
            }
            builder.addAction(actionBuilder.build())
        }
    }

    /** Resolves a button icon from a `resource://type/name` path (0 = no icon). */
    private fun buttonIcon(context: Context, button: Map<String, Any?>): Int {
        val icon = mapUtils.getString(button[Definitions.NOTIFICATION_BUTTON_ICON]) ?: return 0
        if (!icon.startsWith("resource://")) return 0
        val parts = icon.substring("resource://".length).split("/")
        if (parts.size < 2) return 0
        return context.resources.getIdentifier(parts[1], parts[0], context.packageName)
    }

    /**
     * The PendingIntent fired when a button is pressed. `Default` (foreground)
     * buttons launch the app; the others broadcast to [NotificationButtonReceiver]
     * without opening it. Both carry the model JSON + the pressed key.
     */
    private fun buttonPendingIntent(
        context: Context,
        id: Int,
        index: Int,
        key: String,
        model: Map<String, Any?>,
        isForeground: Boolean,
        requireInput: Boolean
    ): PendingIntent {
        val action = Definitions.NOTIFICATION_BUTTON_ACTION_PREFIX + "_" + key
        val requestCode = id * 1000 + index + 1
        return if (isForeground) {
            val launch = context.packageManager
                .getLaunchIntentForPackage(context.packageName) ?: Intent()
            launch.action = action
            launch.putExtra(Definitions.NOTIFICATION_JSON, jsonUtils.toJson(model))
            launch.putExtra(Definitions.NOTIFICATION_BUTTON_KEY_PRESSED, key)
            launch.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP)
            PendingIntent.getActivity(context, requestCode, launch, pendingFlags(requireInput))
        } else {
            val intent = Intent(context, NotificationButtonReceiver::class.java).apply {
                this.action = action
                putExtra(Definitions.NOTIFICATION_JSON, jsonUtils.toJson(model))
                putExtra(Definitions.NOTIFICATION_BUTTON_KEY_PRESSED, key)
            }
            PendingIntent.getBroadcast(context, requestCode, intent, pendingFlags(requireInput))
        }
    }

    /** A button map looked up by its key (null for the content tap / unknown). */
    fun findButton(model: Map<String, Any?>, key: String): Map<String, Any?>? {
        if (key.isEmpty()) return null
        return actionButtons(model).firstOrNull {
            (mapUtils.getString(it[Definitions.NOTIFICATION_BUTTON_KEY]) ?: "") == key
        }
    }

    fun buttonActionType(button: Map<String, Any?>?): String =
        mapUtils.getString(button?.get(Definitions.NOTIFICATION_ACTION_TYPE)) ?: "Default"

    /**
     * Whether a pressed button auto-dismisses its notification. `DismissAction`
     * always dismisses; `KeepOnTop` never; otherwise honor the button's
     * `autoDismissible` flag (default true).
     */
    fun shouldButtonAutoDismiss(button: Map<String, Any?>?): Boolean {
        val actionType = buttonActionType(button)
        if (actionType.endsWith("DismissAction")) return true
        if (actionType.endsWith("KeepOnTop")) return false
        return mapUtils.getBool(button?.get(Definitions.NOTIFICATION_AUTO_DISMISSIBLE)) ?: true
    }

    /**
     * Applies the large icon and (for the BigPicture layout) the big picture,
     * loading bitmaps from any of the four media sources. Runs I/O — call build()
     * off the main thread.
     */
    private fun applyImages(
        context: Context,
        content: Map<String, Any?>,
        builder: NotificationCompat.Builder
    ) {
        val layout = mapUtils.getString(content[Definitions.NOTIFICATION_LAYOUT]) ?: ""
        val bigPicturePath = mapUtils.getString(content[Definitions.NOTIFICATION_BIG_PICTURE])
        val largeIconPath = mapUtils.getString(content[Definitions.NOTIFICATION_LARGE_ICON])
        val isBigPicture = layout.endsWith("BigPicture") && !bigPicturePath.isNullOrEmpty()

        if (isBigPicture) {
            val bigPicture = bitmapUtils.getBitmapFromSource(context, bigPicturePath)
            val largeIcon = if (!largeIconPath.isNullOrEmpty()) {
                bitmapUtils.getBitmapFromSource(context, largeIconPath)
            } else {
                null
            }
            if (largeIcon != null) builder.setLargeIcon(largeIcon)
            if (bigPicture != null) {
                builder.setStyle(
                    NotificationCompat.BigPictureStyle()
                        .bigPicture(bigPicture)
                        .bigLargeIcon(null as android.graphics.Bitmap?)
                )
            }
        } else if (!largeIconPath.isNullOrEmpty()) {
            bitmapUtils.getBitmapFromSource(context, largeIconPath)
                ?.let { builder.setLargeIcon(it) }
        }
    }

    fun notificationId(model: Map<String, Any?>): Int? =
        mapUtils.getInt(contentMap(model)[Definitions.NOTIFICATION_ID])

    @Suppress("UNCHECKED_CAST")
    fun contentMap(model: Map<String, Any?>): Map<String, Any?> =
        (model[Definitions.NOTIFICATION_CONTENT] as? Map<String, Any?>) ?: emptyMap()

    // MARK: - Payload injection (tap / dismiss)

    /** Tap intent: launches the app carrying the full model JSON. */
    private fun contentIntent(context: Context, id: Int, model: Map<String, Any?>): PendingIntent {
        val launch = context.packageManager
            .getLaunchIntentForPackage(context.packageName) ?: Intent()
        launch.action = Definitions.SELECT_NOTIFICATION
        launch.putExtra(Definitions.NOTIFICATION_JSON, jsonUtils.toJson(model))
        launch.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP)
        return PendingIntent.getActivity(context, id, launch, pendingFlags())
    }

    /** Dismiss intent: fired on swipe-away; DismissedNotificationReceiver emits. */
    private fun deleteIntent(context: Context, id: Int, model: Map<String, Any?>): PendingIntent {
        val intent = Intent(context, DismissedNotificationReceiver::class.java).apply {
            action = Definitions.DISMISSED_NOTIFICATION
            putExtra(Definitions.NOTIFICATION_JSON, jsonUtils.toJson(model))
        }
        return PendingIntent.getBroadcast(context, id, intent, pendingFlags())
    }

    /**
     * Mutable intents are required for RemoteInput (text reply) buttons so the
     * system can write the typed text back; everything else stays immutable.
     */
    private fun pendingFlags(mutable: Boolean = false): Int {
        var flags = PendingIntent.FLAG_UPDATE_CURRENT
        if (mutable) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                flags = flags or PendingIntent.FLAG_MUTABLE
            }
            // pre-S: PendingIntents are mutable by default, no flag needed.
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            flags = flags or PendingIntent.FLAG_IMMUTABLE
        }
        return flags
    }

    /** Rebuilds the model map from an intent's NOTIFICATION_JSON extra. */
    fun notificationModel(intent: Intent): Map<String, Any?>? {
        val json = intent.getStringExtra(Definitions.NOTIFICATION_JSON) ?: return null
        return jsonUtils.fromJson(json)
    }

    // MARK: - Event registration (stamps date + lifecycle on the content)

    fun registerCreatedEvent(content: Map<String, Any?>, lifeCycle: String): Map<String, Any?> =
        content + mapOf(
            Definitions.NOTIFICATION_CREATED_DATE to now(),
            Definitions.NOTIFICATION_CREATED_LIFECYCLE to lifeCycle,
            Definitions.NOTIFICATION_CREATED_SOURCE to "Local"
        )

    fun registerDisplayedEvent(content: Map<String, Any?>, lifeCycle: String): Map<String, Any?> =
        content + mapOf(
            Definitions.NOTIFICATION_DISPLAYED_DATE to now(),
            Definitions.NOTIFICATION_DISPLAYED_LIFECYCLE to lifeCycle
        )

    fun registerActionEvent(
        content: Map<String, Any?>,
        lifeCycle: String,
        actionType: String = "Default"
    ): Map<String, Any?> =
        content + mapOf(
            Definitions.NOTIFICATION_ACTION_DATE to now(),
            Definitions.NOTIFICATION_ACTION_LIFECYCLE to lifeCycle,
            Definitions.NOTIFICATION_ACTION_TYPE to actionType
        )

    fun registerDismissedEvent(content: Map<String, Any?>, lifeCycle: String): Map<String, Any?> =
        content + mapOf(
            Definitions.NOTIFICATION_DISMISSED_DATE to now(),
            Definitions.NOTIFICATION_ACTION_LIFECYCLE to lifeCycle
        )

    // MARK: - Dismiss concepts

    fun readId(content: Map<String, Any?>): Int? =
        mapUtils.getInt(content[Definitions.NOTIFICATION_ID])

    /** A `DismissAction` tap dismisses + fires the dismiss event. */
    fun isDismissAction(content: Map<String, Any?>): Boolean =
        (mapUtils.getString(content[Definitions.NOTIFICATION_ACTION_TYPE]) ?: "")
            .endsWith("DismissAction")

    /**
     * Whether tapping the notification auto-dismisses it (Android `setAutoCancel`).
     * DismissAction always dismisses; KeepOnTop never; otherwise honor the
     * `autoDismissible` flag (default true).
     */
    fun shouldAutoDismiss(content: Map<String, Any?>): Boolean {
        val actionType =
            mapUtils.getString(content[Definitions.NOTIFICATION_ACTION_TYPE]) ?: ""
        if (actionType.endsWith("DismissAction")) return true
        if (actionType.endsWith("KeepOnTop")) return false
        return mapUtils.getBool(content[Definitions.NOTIFICATION_AUTO_DISMISSIBLE]) ?: true
    }

    private fun now(): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US)
        formatter.timeZone = TimeZone.getTimeZone("UTC")
        return formatter.format(Date())
    }
}

File: native/apple/Sources/IosAwnCore/NotificationBuilder.swift
import Foundation
import UserNotifications

/// Builds notifications from a serialized `NotificationModel`, injects the full
/// model into the delivered notification so it can be recovered later, and
/// stamps the lifecycle events (created/displayed/pressed/dismissed).
///
/// Created via a factory (an instance with injected helpers — not static utility
/// methods), mirroring the original IosAwnCore `NotificationBuilder`.
public final class NotificationBuilder {

    public static func newInstance() -> NotificationBuilder {
        return NotificationBuilder()
    }
    private init() {}

    private let stringUtils = StringUtils.shared
    private let mapUtils = MapUtils.shared
    private let jsonUtils = JsonUtils.shared

    // MARK: - Build

    /// Builds the content for a notification from a serialized model, injecting
    /// the full model JSON so it can be recovered on display/tap/dismiss.
    /// Returns the notification id and the built content, or nil if invalid.
    public func createNotificationContent(
        fromModel model: [String: Any]
    ) -> (id: Int, content: UNMutableNotificationContent)? {
        let content = contentMap(fromModel: model)
        guard let id = mapUtils.getInt(content[Definitions.NOTIFICATION_ID]) else {
            return nil
        }

        let unContent = UNMutableNotificationContent()
        let title = mapUtils.getString(content[Definitions.NOTIFICATION_TITLE])
        if !stringUtils.isNullOrEmpty(title) { unContent.title = title! }
        let body = mapUtils.getString(content[Definitions.NOTIFICATION_BODY])
        if !stringUtils.isNullOrEmpty(body) { unContent.body = body! }
        unContent.sound = .default
        unContent.categoryIdentifier = categoryIdentifier(forModel: model)

        applyImages(content, to: unContent)
        setUserInfoContent(model: model, content: unContent)
        return (id, unContent)
    }

    // MARK: - Action buttons

    /// The serialized action buttons of a model (the `actionButtons` list).
    public func actionButtons(fromModel model: [String: Any]) -> [[String: Any]] {
        return (model[Definitions.NOTIFICATION_BUTTONS] as? [[String: Any]]) ?? []
    }

    /// The category identifier a notification should use: `DEFAULT` when it has
    /// no buttons, otherwise a deterministic id derived from the button keys so
    /// the same set of buttons reuses the same registered category.
    public func categoryIdentifier(forModel model: [String: Any]) -> String {
        let keys = actionButtons(fromModel: model)
            .compactMap { mapUtils.getString($0[Definitions.NOTIFICATION_BUTTON_KEY]) }
            .filter { !$0.isEmpty }
        return keys.isEmpty ? Definitions.DEFAULT_CATEGORY_IDENTIFIER : keys.joined(separator: ",")
    }

    /// Builds the `UNNotificationCategory` (with its actions) for a model's
    /// buttons, or nil when there are none. The caller registers it.
    public func actionCategory(forModel model: [String: Any]) -> UNNotificationCategory? {
        let buttons = actionButtons(fromModel: model)
        var actions: [UNNotificationAction] = []
        for button in buttons {
            guard let key = mapUtils.getString(button[Definitions.NOTIFICATION_BUTTON_KEY]),
                  !key.isEmpty else { continue }
            let label = mapUtils.getString(button[Definitions.NOTIFICATION_BUTTON_LABEL]) ?? key
            let actionType =
                mapUtils.getString(button[Definitions.NOTIFICATION_ACTION_TYPE]) ?? "Default"

            var options: UNNotificationActionOptions = []
            if actionType.hasSuffix("Default") { options.insert(.foreground) }
            if mapUtils.getBool(button[Definitions.NOTIFICATION_IS_DANGEROUS_OPTION]) == true {
                options.insert(.destructive)
            }
            if mapUtils.getBool(button[Definitions.NOTIFICATION_AUTHENTICATION_REQUIRED]) == true {
                options.insert(.authenticationRequired)
            }

            if mapUtils.getBool(button[Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT]) == true {
                actions.append(UNTextInputNotificationAction(
                    identifier: key, title: label, options: options))
            } else {
                actions.append(UNNotificationAction(
                    identifier: key, title: label, options: options))
            }
        }
        if actions.isEmpty { return nil }
        return UNNotificationCategory(
            identifier: categoryIdentifier(forModel: model),
            actions: actions,
            intentIdentifiers: [],
            options: [.customDismissAction]
        )
    }

    /// Finds a button map by its key (nil for the content tap / unknown key).
    public func findButton(inModel model: [String: Any], key: String) -> [String: Any]? {
        if key.isEmpty { return nil }
        return actionButtons(fromModel: model).first {
            (mapUtils.getString($0[Definitions.NOTIFICATION_BUTTON_KEY]) ?? "") == key
        }
    }

    /// Attaches the notification image. iOS shows a single attachment, so prefer
    /// the big picture and fall back to the large icon.
    private func applyImages(_ content: [String: Any], to unContent: UNMutableNotificationContent) {
        let bigPicture = mapUtils.getString(content[Definitions.NOTIFICATION_BIG_PICTURE])
        let largeIcon = mapUtils.getString(content[Definitions.NOTIFICATION_LARGE_ICON])
        let source = !(bigPicture?.isEmpty ?? true) ? bigPicture : largeIcon
        if let attachment = bitmapAttachment(from: source) {
            unContent.attachments = [attachment]
        }
    }

    private func bitmapAttachment(from source: String?) -> UNNotificationAttachment? {
        #if canImport(UIKit)
        guard let source = source, !source.isEmpty,
              let image = BitmapUtils.shared.getBitmapFromSource(source),
              let data = image.pngData()
        else { return nil }

        // Each call writes into its own unique temp subfolder (like the original
        // core's `globallyUniqueString` subfolder), so concurrent builds never
        // share a path. The attachment identifier is unique too.
        let uniqueName = UUID().uuidString
        let directory = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent(uniqueName, isDirectory: true)
        do {
            try FileManager.default.createDirectory(
                at: directory, withIntermediateDirectories: true)
            let fileURL = directory.appendingPathComponent(uniqueName + ".png")
            try data.write(to: fileURL)
            return try UNNotificationAttachment(
                identifier: uniqueName + ".png", url: fileURL, options: nil)
        } catch {
            return nil
        }
        #else
        return nil
        #endif
    }

    /// Stores the full serialized model into userInfo for later recovery.
    private func setUserInfoContent(
        model: [String: Any],
        content: UNMutableNotificationContent
    ) {
        if let json = jsonUtils.toJson(model) {
            content.userInfo[Definitions.NOTIFICATION_JSON] = json
        }
    }

    // MARK: - Recover

    /// Rebuilds the model map from a delivered notification's userInfo.
    public func notificationModel(fromUserInfo userInfo: [AnyHashable: Any]) -> [String: Any]? {
        guard let json = userInfo[Definitions.NOTIFICATION_JSON] as? String else {
            return nil
        }
        return jsonUtils.fromJson(json)
    }

    /// The content sub-map of a serialized model.
    public func contentMap(fromModel model: [String: Any]) -> [String: Any] {
        return (model[Definitions.NOTIFICATION_CONTENT] as? [String: Any]) ?? [:]
    }

    // MARK: - Event registration (stamps date + lifecycle on the content)

    public func registerCreatedEvent(_ content: [String: Any], lifeCycle: String) -> [String: Any] {
        var map = content
        map[Definitions.NOTIFICATION_CREATED_DATE] = now()
        map[Definitions.NOTIFICATION_CREATED_LIFECYCLE] = lifeCycle
        map[Definitions.NOTIFICATION_CREATED_SOURCE] = "Local"
        return map
    }

    public func registerDisplayedEvent(_ content: [String: Any], lifeCycle: String) -> [String: Any] {
        var map = content
        map[Definitions.NOTIFICATION_DISPLAYED_DATE] = now()
        map[Definitions.NOTIFICATION_DISPLAYED_LIFECYCLE] = lifeCycle
        return map
    }

    public func registerActionEvent(
        _ content: [String: Any],
        lifeCycle: String,
        actionType: String = "Default"
    ) -> [String: Any] {
        var map = content
        map[Definitions.NOTIFICATION_ACTION_DATE] = now()
        map[Definitions.NOTIFICATION_ACTION_LIFECYCLE] = lifeCycle
        map[Definitions.NOTIFICATION_ACTION_TYPE] = actionType
        return map
    }

    public func registerDismissedEvent(_ content: [String: Any], lifeCycle: String) -> [String: Any] {
        var map = content
        map[Definitions.NOTIFICATION_DISMISSED_DATE] = now()
        map[Definitions.NOTIFICATION_ACTION_LIFECYCLE] = lifeCycle
        return map
    }

    // MARK: - Dismiss concepts

    func readId(_ content: [String: Any]) -> Int? {
        return mapUtils.getInt(content[Definitions.NOTIFICATION_ID])
    }

    /// A `DismissAction` notification/button behaves like a user dismissal: it
    /// dismisses the notification and fires the dismiss event (ignoring
    /// autoDismissible).
    func isDismissAction(_ content: [String: Any]) -> Bool {
        let actionType =
            mapUtils.getString(content[Definitions.NOTIFICATION_ACTION_TYPE]) ?? ""
        return actionType.hasSuffix("DismissAction")
    }

    /// The actionType of a pressed button (or "Default" when the content itself
    /// was tapped / the button is unknown).
    func buttonActionType(_ button: [String: Any]?) -> String {
        return mapUtils.getString(button?[Definitions.NOTIFICATION_ACTION_TYPE] ?? "")
            ?? "Default"
    }

    /// Whether a pressed button should auto-dismiss its notification.
    /// `DismissAction` always dismisses; `KeepOnTop` never; otherwise honor the
    /// button's `autoDismissible` flag (default true).
    func shouldButtonAutoDismiss(_ button: [String: Any]?) -> Bool {
        let actionType = buttonActionType(button)
        if actionType.hasSuffix("DismissAction") { return true }
        if actionType.hasSuffix("KeepOnTop") { return false }
        return mapUtils.getBool(button?[Definitions.NOTIFICATION_AUTO_DISMISSIBLE] ?? "")
            ?? true
    }

    private func now() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        formatter.timeZone = TimeZone(identifier: "UTC")
        return formatter.string(from: Date())
    }
}
